### PR TITLE
NNI Engine: Add Filtering Scheme

### LINKED ISSUE #429: NNI Engine: Filtering Scheme
NNI Engine needs a scheme for determining which NNI to accept and reject.

### PR DESCRIPTION
**Pull request by davidrich27**
## Description

- Adds Filtering Schemes for accepting or rejecting proposed NNIs.
   - Added std::function members to NNI Engine, which can be modified to select different filtering schemes, or user make a custom scheme by passing lambdas.
   - Added Schemes:  
     - None (accepts all NNIs)
     - Cutoff (accepts NNIs with likelihood greater than cutoff threshold).
     - Hill-climbing / Loss threshold (accepts NNIs which have higher likelihood than their Pre-NNIs present in DAG, or do not fall below a certain threshold).
- Added NNI Engine to Pybito.

Closes #429 

## Tests

`TEST_CASE("NNI Engine: Complete Run NNI Engine with No Filter")`:  Tests that running NNI Engine with no filter results in a complete subsplitDAG.

`TEST_CASE("NNI Engine: Run Hillclimbing")`: Tests that hillclimbing only accepts NNIs with higher likelihood.

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [x] Issue branch has been squashed and rebased on main branch
* [ ] GitHub CI build on PR branch completed successfully

### CODE REVIEW

File: test/test_bito.py [hunk 1]
```diff
@@ -180,6 +182,17 @@ def test_gp_instance():
...
+    nni_engine = inst.get_nni_engine()
+    nni_engine.run_init()
+    adj_nni_count = nni_engine.get_adjacent_nni_count()
+    nni_engine.run()
+
+    dag = inst.get_dag()
+    node_count = dag.get_node_count()
+    edge_count = dag.get_edge_count()
+    print("nni_stats: ", adj_nni_count, node_count, edge_count)
+
```
Comment (matsen) [inline test/test_bito.py:195]: Is this mostly a demo vs a test? Do we want to have test conditions? Also, isn't the pass just below superfluous?

File: src/nni_engine.hpp [hunk 2]
```diff
@@ -30,32 +30,53 @@ class NNIEngine {
...
+  size_t GetSweepCount() const { return sweep_count_; }
+
+  // ** Setters
+
+  // Set whether to track accepted past NNIs.
+  void SetKeepAcceptedNNIs(const bool keep_accepted_past_nnis) {
+    keep_accepted_past_nnis_ = keep_accepted_past_nnis;
+  }
+  // Set whether to track rejected past NNIs.
+  void SetKeepRejecteddNNIs(const bool keep_rejected_past_nnis) {
```
Comment (matsen) [inline src/nni_engine.hpp:73]:
```suggestion
  void SetKeepRejectedNNIs(const bool keep_rejected_past_nnis) {
```

**Review by matsen — COMMENTED** [review]
Two silly things

File: test/test_bito.py (hunk 1 above)
Comment (davidrich27) [reply test/test_bito.py:195]: Yes, this is just a demo.

File: src/gp_doctest.cpp [hunk 3]
```diff
@@ -1562,3 +1562,147 @@ TEST_CASE("NNI Engine: NNI Likelihoods") {
     nni_count++;
   }
 }
+
+// Build set all valid node subsplits present in a complete SubsplitDAG.
```
Comment (matsen) [inline src/gp_doctest.cpp:1566]: Let's get a bit more explanation here.

File: src/gp_doctest.cpp [hunk 4]
```diff
@@ -1562,3 +1562,147 @@ TEST_CASE("NNI Engine: NNI Likelihoods") {
...
+  // Outgroup.
+  CompleteDAGSubsplits(subsplits, remaining - 1, assignment + "0", first_left_subsplit);
+  // Left Subsplit.
+  CompleteDAGSubsplits(subsplits, remaining - 1, assignment + "1", true);
+  // Right Subsplit (the first non-outgroup must be assigned to the Left Subpslit).
+  if (first_left_subsplit) {
+    CompleteDAGSubsplits(subsplits, remaining - 1, assignment + "2", true);
+  }
+}
+// Build set all valid edge PCSPs present in a complete SubsplitDAG.
```
Comment (matsen) [inline src/gp_doctest.cpp:1611]: I think we could give it an arbitrary collection of bitsets, no?

File: src/gp_doctest.cpp [hunk 5]
```diff
@@ -1562,3 +1562,147 @@ TEST_CASE("NNI Engine: NNI Likelihoods") {
...
+  std::cout << "[BEGIN] HILL_CLIMBING" << std::endl;
+  const std::string fasta_path = "data/six_taxon.fasta";
+  const std::string newick_path = "data/six_taxon_rooted_simple.nwk";
+  auto inst =
+      GPInstanceOfFiles(fasta_path, newick_path, "_ignore/mmapped_plv_graft.data");
+  auto& dag = inst.GetDAG();
+  inst.MakeNNIEngine();
+  auto& nni_engine = inst.GetNNIEngine();
+  auto& graft_dag = nni_engine.GetGraftDAG();
+  // Set NNI Engine to have no filter.
```
Comment (matsen) [inline src/gp_doctest.cpp:1688]: Out of date

File: src/gp_doctest.cpp [hunk 6]
```diff
@@ -1562,3 +1562,147 @@ TEST_CASE("NNI Engine: NNI Likelihoods") {
...
+    std::cout << "AdjacentCount: " << nni_engine.GetAdjacentNNICount() << std::endl;
+    nni_engine.RunMainLoop();
+    nni_engine.RunPostLoop();
+  }
+
+  for (const auto& nni : nni_engine.GetPastAcceptedNNIs()) {
+    const auto prenni_likelihood = nni_engine.GetBestPreNNILikelihood(nni);
+    const auto postnni_likelihood = nni_engine.GetPostNNILikelihood(nni);
+
+    // CHECK_MESSAGE(prenni_likelihood < postnni_likelihood,
```
Comment (matsen) [inline src/gp_doctest.cpp:1704]: fix

File: src/nni_engine.hpp [hunk 7]
```diff
@@ -68,33 +89,71 @@ class NNIEngine {
   void RunMainLoop();
   void RunPostLoop();
 
-  // ** Filter Functions
-
-  // Initialization step for filter before beginning
+  // ** Filtering Functions
+  // Initialization step for filter before beginning engine run.
```
Comment (matsen) [inline src/nni_engine.hpp:93]: @ognian- can you have a chat with @davidrich27 about these definitions just below? The overhead of std::function is _definitely_ not a problem here (vs templates) but if we want to have some code standard about how to handle this let's talk.

File: src/nni_engine.hpp [hunk 8]
```diff
@@ -68,33 +89,71 @@ class NNIEngine {
...
-  using StaticFilterEvaluateFunction =
-      std::function<double(NNIEngine &, GPEngine &, GraftDAG &, const NNIOperation &)>;
   // Function template for processing an adjacent NNI to be accepted or rejected.
   using StaticFilterProcessFunction =
       std::function<bool(NNIEngine &, GPEngine &, GraftDAG &, const NNIOperation &)>;
 
-  // StaticFilterEvaluateFunction: dummy function that assigns all score of 0.0.
-  static double NoEvaluate(NNIEngine &this_nni_engine, GPEngine &this_gp_engine,
-                           GraftDAG &this_graft_dag, const NNIOperation &nni);
+  // StaticFilterInitFunction: Init function that does nothing.
```
Comment (matsen) [inline src/nni_engine.hpp:103]:
```suggestion
  // This StaticFilterInitFunction does nothing.
```

File: src/nni_engine.hpp [hunk 9]
```diff
@@ -68,33 +89,71 @@ class NNIEngine {
...
-  // Pre-NNI in the DAG.
-  static double EstimateNNILikelihoodViaPreNNI(NNIEngine &this_nni_engine,
-                                               GPEngine &this_gp_engine,
-                                               GraftDAG &this_graft_dag,
-                                               const NNIOperation &nni);
+                       GraftDAG &this_graft_dag, const NNIOperation &nni) {
+    return accept;
+  }
+
+  // Helper: Finds best pre-NNI likelihood in DAG for given NNI.
```
Comment (matsen) [inline src/nni_engine.hpp:116]: To cpp, please

Comment (matsen) [reply src/nni_engine.hpp:116]: Explain that the given nni could arise from multiple nnis from PCSPs already in the dag; this function gets the best one of those.

Comment (matsen) [reply src/nni_engine.hpp:116]: Could mark for future consideration on the next round of PR work.

File: src/nni_engine.hpp [hunk 10]
```diff
@@ -68,33 +89,71 @@ class NNIEngine {
...
+          prenni_edge_idx = edge_idx;
+        }
+      }
+    }
+    return prenni_likelihood;
+  }
+  // Helper: Get NNI likelihood.
+  double GetPostNNILikelihood(const NNIOperation &nni) { return scored_nnis_[nni]; }
+
+  //
```
Comment (matsen) [inline src/nni_engine.hpp:139]: comment

File: src/nni_engine.hpp [hunk 11]
```diff
@@ -237,15 +332,15 @@ class NNIEngine {
   double GetScoreForNNI(const NNIOperation &nni) const;
 
   // Update adjacent NNIs at end of current sweep. If not re-evaluating rejected NNIs,
-  // then current adjacent NNIs are removed. Then NNIs that are adjacent to last sweep's
-  // accepted NNIs are added.
-  void UpdateAdjacentNNIs(const bool reevaluate_rejected_nnis = false);
+  // then current adjacent NNIs are removed. Then NNIs that are adjacent to last
+  // sweep's accepted NNIs are added.
+  void UpdateAdjacentNNIs(std::optional<bool> reevaluate_rejected_nnis = std::nullopt);
```
Comment (matsen) [inline src/nni_engine.hpp:337]: let's keep it simple

File: src/pybito.cpp [hunk 12]
```diff
@@ -624,18 +626,44 @@ PYBIND11_MODULE(bito, m) {
...
-          R"raw(Find all adjacent NNIs of DAG.)raw")
-      .def(
-          "adjacent_nni_count",
-          [](GPInstance &self) { self.GetNNIEngine().GetAdjacentNNICount(); },
-          R"raw(Get number of adjacent NNIs to current DAG.)raw");
+      .def("get_nni_engine", &GPInstance::GetNNIEngine,
+           py::return_value_policy::reference, R"raw(Initialize NNI Engine.)raw");
+
+  // CLASS
+  // SubsplitDAG
```
Comment (matsen) [inline src/pybito.cpp:635]: check

File: src/subsplit_dag.cpp [hunk 13]
```diff
@@ -285,6 +285,85 @@ SizeBoolVectorMap SubsplitDAG::BuildEdgeIdxToPCSPBoolVectorMap() const {
...
+  }
+  std::cout << "clade_counts: " << unique_count << " " << clades.size() << std::endl;
+  return clades;
+}
+
+template <typename KeyType, typename DataType>
+void VectorMapInsertOrAppend(std::map<KeyType, std::vector<DataType>> &map, KeyType key,
+                             DataType data) {
+  if (map.find(key) == map.end()) {
+    map.insert({key, SizeVector()});
```
Comment (matsen) [inline src/subsplit_dag.cpp:348]: This doesn't seem quite right

**Review by matsen — CHANGES_REQUESTED** [review]
🚄

File: src/nni_engine.hpp (hunk 7 above)
Comment (ognian-) [reply src/nni_engine.hpp:93]: I think using std::function is a great option, it avoids templating and is readable.

File: src/nni_engine.hpp (hunk 2 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Fixed.

File: src/gp_doctest.cpp (hunk 3 above)
Comment (davidrich27) [reply src/gp_doctest.cpp]: Added explanation.

File: src/gp_doctest.cpp (hunk 4 above)
Comment (davidrich27) [reply src/gp_doctest.cpp]: Fixed comment.

File: src/gp_doctest.cpp (hunk 5 above)
Comment (davidrich27) [reply src/gp_doctest.cpp]: Removed.

File: src/gp_doctest.cpp (hunk 6 above)
Comment (davidrich27) [reply src/gp_doctest.cpp]: Fixed.

File: src/nni_engine.hpp (hunk 9 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Moved and added notes.

File: src/nni_engine.hpp (hunk 10 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Fixed.

File: src/nni_engine.hpp (hunk 11 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Removed.  Option now set though class member.

File: src/pybito.cpp (hunk 12 above)
Comment (davidrich27) [reply src/pybito.cpp]: Switched comment to `GPDAG`.

File: src/subsplit_dag.cpp (hunk 13 above)
Comment (davidrich27) [reply src/subsplit_dag.cpp]: Fixed.

File: src/nni_engine.cpp [hunk 14]
```diff
@@ -33,43 +33,88 @@ void NNIEngine::RunInit() {
...
-bool NNIEngine::NoFilter(NNIEngine &this_nni_engine, GPEngine &this_gp_engine,
-                         GraftDAG &this_graft_dag, const NNIOperation &nni) {
-  return true;
+bool NNIEngine::FilterPostNNIBetterThanPreNNI(NNIEngine &this_nni_engine,
+                                              GPEngine &this_gp_engine,
+                                              GraftDAG &this_graft_dag,
+                                              const NNIOperation &nni) {
+  const double postnni_likelihood = this_nni_engine.GetPostNNILikelihood(nni);
+  const double prenni_likelihood = this_nni_engine.GetBestPreNNILikelihood(nni);
+  const double diff = postnni_likelihood - prenni_likelihood;
```
Comment (matsen) [inline src/nni_engine.cpp:115]: Wouldn't this be clearer and simpler if we just had 
```
return postnni_likelihood > prenni_likelihood;
```
?

**Review by matsen — APPROVED** [review]
I made one tiny suggestion, but this is ready to ship. Thanks!

File: src/nni_engine.cpp (hunk 14 above)
Comment (davidrich27) [reply src/nni_engine.cpp:115]: Yep!  Changed.

**Pull request merged**